### PR TITLE
[VitisAI]Update VitisAI EP to be compatible with VitisAI 3.5

### PR DESCRIPTION
**Pull request by wcy123**
### Description

Originally VitisAI EP only works with old version of VitisAI release. 

### Motivation and Context

Update VitisAI EP so that it works together with the current VitisiAI 3.5 and further version of VitisAI. We try our best to make it forward compatible.

### CODE REVIEW

**Comment by wcy123** [comment]
> @wcy123 the command you issued was incorrect. Please try again.
> 
> Examples are:
> 
> ```
> @microsoft-github-policy-service agree
> ```
> 
> and
> 
> ```
> @microsoft-github-policy-service agree company="your company"
> ```

@microsoft-github-policy-service agree company="AMD"

File: include/onnxruntime/core/providers/vitisai/vitisai_provider_factory.h [hunk 1]
```diff
@@ -10,9 +10,12 @@ extern "C" {
...
  */
-ORT_API_STATUS(OrtSessionOptionsAppendExecutionProvider_VITISAI, _In_ OrtSessionOptions* options,
-               const char* backend_type, int device_id, const char* export_runtime_module,
-               const char* load_runtime_module);
+ORT_API_STATUS(OrtSessionOptionsAppendExecutionProvider_VITISAI, _In_ OrtSessionOptions* options, _In_ const char* opt_str);
+
+/**
+ *  initialize the vitis ai execution providers.
+ */
+ORT_API_STATUS(OrtExecutionProviderInitialize_VITISAI);
```
Comment (jywu-msft) [inline include/onnxruntime/core/providers/vitisai/vitisai_provider_factory.h:18]: why is this needed? having this api would be inconsistent with all other EP's.

File: cmake/onnxruntime_providers.cmake [hunk 2]
```diff
@@ -768,19 +768,46 @@ if (onnxruntime_USE_TENSORRT)
 endif()
 
 if (onnxruntime_USE_VITISAI)
-  file(GLOB_RECURSE onnxruntime_providers_vitisai_cc_srcs CONFIGURE_DEPENDS
-    "${ONNXRUNTIME_ROOT}/core/providers/vitisai/*.h"
+  if ("${GIT_COMMIT_ID}" STREQUAL "")
+  execute_process(
+    WORKING_DIRECTORY ${CMAKE_CURRENT_SOURCE_DIR}
+    COMMAND git rev-parse HEAD
```
Comment (jywu-msft) [inline cmake/onnxruntime_providers.cmake:774]: why is GIT_COMMIT_ID needed specifically for vitisai?

File: include/onnxruntime/core/providers/vitisai/vitisai_provider_factory.h [hunk 3]
```diff
@@ -10,9 +10,12 @@ extern "C" {
 /**
  * \param use_arena zero: false. non-zero: true.
  */
-ORT_API_STATUS(OrtSessionOptionsAppendExecutionProvider_VITISAI, _In_ OrtSessionOptions* options,
-               const char* backend_type, int device_id, const char* export_runtime_module,
-               const char* load_runtime_module);
+ORT_API_STATUS(OrtSessionOptionsAppendExecutionProvider_VITISAI, _In_ OrtSessionOptions* options, _In_ const char* opt_str);
```
Comment (jywu-msft) [inline include/onnxruntime/core/providers/vitisai/vitisai_provider_factory.h:13]: could https://github.com/microsoft/onnxruntime/blob/c6074f3a4b09cb44b46ca4422717deb067015bfa/include/onnxruntime/core/session/onnxruntime_c_api.h#L3527-L3531 be used instead, then one wouldn't need to introduce a new api.

**Comment by wcy123** [comment]
发自我的 iPhone

在 2023年4月27日，07:54，George Wu ***@***.***> 写道：

﻿
Caution: This message originated from an External Source. Use proper caution when opening attachments, clicking links, or responding.


@jywu-msft commented on this pull request.

________________________________

In include/onnxruntime/core/providers/vitisai/vitisai_provider_factory.h<https://github.com/microsoft/onnxruntime/pull/15673#discussion_r1178498066>:

> @@ -10,9 +10,12 @@ extern "C" {
 /**
  * \param use_arena zero: false. non-zero: true.
  */
-ORT_API_STATUS(OrtSessionOptionsAppendExecutionProvider_VITISAI, _In_ OrtSessionOptions* options,
-               const char* backend_type, int device_id, const char* export_runtime_module,
-               const char* load_runtime_module);
+ORT_API_STATUS(OrtSessionOptionsAppendExecutionProvider_VITISAI, _In_ OrtSessionOptions* options, _In_ const char* opt_str);


could https://github.com/microsoft/onnxruntime/blob/c6074f3a4b09cb44b46ca4422717deb067015bfa/include/onnxruntime/core/session/onnxruntime_c_api.h#L3527-L3531 be used instead, then one wouldn't need to introduce a new api.

**Comment by wcy123** [comment]
Mingyue please help, I am on annual leave for today.

发自我的 iPhone

在 2023年4月27日，07:54，George Wu ***@***.***> 写道：

﻿
Caution: This message originated from an External Source. Use proper caution when opening attachments, clicking links, or responding.


@jywu-msft commented on this pull request.

________________________________

In include/onnxruntime/core/providers/vitisai/vitisai_provider_factory.h<https://github.com/microsoft/onnxruntime/pull/15673#discussion_r1178498066>:

> @@ -10,9 +10,12 @@ extern "C" {
 /**
  * \param use_arena zero: false. non-zero: true.
  */
-ORT_API_STATUS(OrtSessionOptionsAppendExecutionProvider_VITISAI, _In_ OrtSessionOptions* options,
-               const char* backend_type, int device_id, const char* export_runtime_module,
-               const char* load_runtime_module);
+ORT_API_STATUS(OrtSessionOptionsAppendExecutionProvider_VITISAI, _In_ OrtSessionOptions* options, _In_ const char* opt_str);


could https://github.com/microsoft/onnxruntime/blob/c6074f3a4b09cb44b46ca4422717deb067015bfa/include/onnxruntime/core/session/onnxruntime_c_api.h#L3527-L3531 be used instead, then one wouldn't need to introduce a new api.

File: cmake/onnxruntime_providers.cmake (hunk 2 above)
Comment (mingyueliuh) [reply cmake/onnxruntime_providers.cmake:774]: Used to obtain the version of ONNXRuntime Vitis AI EP in the runtime phase, which is very useful for reproducing issues in the troubleshooting  phase.

File: include/onnxruntime/core/providers/vitisai/vitisai_provider_factory.h (hunk 3 above)
Comment (wcy123) [reply include/onnxruntime/core/providers/vitisai/vitisai_provider_factory.h:13]: We will update it soon.

File: onnxruntime/core/providers/vitisai/imp/tensor_proto.cpp [hunk 4]
```diff
@@ -0,0 +1,119 @@
+// Copyright (c) 2023 Advanced Micro Devices, Inc. All rights reserved.
```
Comment (jywu-msft) [inline onnxruntime/core/providers/vitisai/imp/tensor_proto.cpp:1]: https://github.com/microsoft/onnxruntime/blob/main/onnxruntime/core/framework/tensorprotoutils.h

File: onnxruntime/core/providers/vitisai/imp/attr_proto.cpp [hunk 5]
```diff
@@ -0,0 +1,127 @@
+// Copyright (c) 2023 Advanced Micro Devices, Inc. All rights reserved.
```
Comment (jywu-msft) [inline onnxruntime/core/providers/vitisai/imp/attr_proto.cpp:1]: https://github.com/microsoft/onnxruntime/blob/main/onnxruntime/core/graph/node_attr_utils.cc

File: cmake/onnxruntime_providers.cmake [hunk 6]
```diff
@@ -768,19 +768,46 @@ if (onnxruntime_USE_TENSORRT)
...
   )
-
   source_group(TREE ${ONNXRUNTIME_ROOT}/core FILES ${onnxruntime_providers_vitisai_cc_srcs})
   onnxruntime_add_static_library(onnxruntime_providers_vitisai ${onnxruntime_providers_vitisai_cc_srcs})
-  onnxruntime_add_include_to_target(onnxruntime_providers_vitisai
-    onnxruntime_common onnxruntime_framework onnx onnx_proto ${PROTOBUF_LIB} flatbuffers::flatbuffers Boost::mp11 safeint_interface
-  )
-  add_dependencies(onnxruntime_providers_vitisai ${onnxruntime_EXTERNAL_DEPENDENCIES})
+  onnxruntime_add_include_to_target(onnxruntime_providers_vitisai onnxruntime_common onnxruntime_framework onnx onnx_proto)
+  add_library(onnxruntime_vitisai_ep SHARED ${ONNXRUNTIME_ROOT}/core/providers/vitisai/onnxruntime_vitisai_ep_stub.cpp)
```
Comment (jywu-msft) [inline cmake/onnxruntime_providers.cmake:788]: you can't use onnxruntime_add_shared_library_module() ?

**Review by pranavsharma — COMMENTED** [review]
C++ source files should have the extension .h or .cc to be consistent with the rest of the codebase.

**Comment by skottmckay** [comment]
Can you please add a readme.md to onnxruntime\core\providers\vitisai that provides contact/ownership information for the EP.

**Comment by wcy123** [comment]
> general question. i notice there's a lot of utility functions added for manipulating graph/onnx proto etc.
> have you looked into https://github.com/microsoft/onnxruntime/tree/main/onnxruntime/core/graph to see what can be reused?
> since vitis EP seems to be built as a shared lib, you must use the bridge api to leverage ort core/framework modules.
> see https://github.com/microsoft/onnxruntime/blob/main/onnxruntime/core/providers/shared_library/provider_interfaces.h
> for what interfaces have been bridged. if there are missing ones, they can be added.

Hi @jywu-msft 

      This project started one year ago, at that time, we didn't see `shared_library/provider_interfaces.h`. I agree that `shared_library/provider_interfaces.h` is much similar to these utility functions now, however it is too later for us to adapt the new interfaces.  Development iteration was also an issue because it takes very long time for the internal legal scan. 

      Now the legal scan is done, we can submit our code to the upstream, and we can start quick development interaction with the upstream.  So, can we manage and plan to use `shared_library/provider_interfaces.h` in the next release?

      For now, it is not possible to update source codes to use `shared_library/provider_interfaces.h` within such small release window.

Br
Wang Chunye

File: cmake/onnxruntime_providers.cmake (hunk 6 above)
Comment (wcy123) [reply cmake/onnxruntime_providers.cmake]: in 0b94a07, we use `onnxruntime_add_shared_library`. However, `onnxruntime_vitisai_ep` must be a shared library and it is replaced by `onnxruntime_vitisai_ep.dll` in VitisAI 3.5 or VitisAI4.0 releases.

The real `onnxruntime_vitisai_ep.dll` is not ready for open source yet.

File: include/onnxruntime/core/providers/vitisai/vitisai_provider_factory.h (hunk 3 above)
Comment (wcy123) [reply include/onnxruntime/core/providers/vitisai/vitisai_provider_factory.h]: > We will update it soon.
77db8d25a1d2b16ee07598fbbf99186691708256 fix the issue.

File: onnxruntime/core/providers/vitisai/vitisai_execution_provider.cc [hunk 7]
```diff
@@ -1,300 +1,145 @@
...
-      if (cluster_idx.find(xl->subgraph) == cluster_idx.end()) {
-        cluster_idx[xl->subgraph] = cur_idx;
-        std::vector<NodeIndex> new_cluster;
-        clusters.push_back(new_cluster);
-      }
-    }
+struct MyCustomOpKernel : OpKernel {
+  MyCustomOpKernel(const OpKernelInfo& info, const OrtCustomOp& op) : OpKernel(info), op_(op) {
+    op_kernel_ = op_.CreateKernel(&op_, OrtGetApiBase()->GetApi(op_.version),
+                                  reinterpret_cast<const OrtKernelInfo*>(&info));
```
Comment (skottmckay) [inline onnxruntime/core/providers/vitisai/vitisai_execution_provider.cc:41]: What's the reason for using the external ORT API for internal EP code?

File: onnxruntime/core/providers/vitisai/vitisai_execution_provider.h [hunk 8]
```diff
@@ -1,20 +1,27 @@
...
 namespace onnxruntime {
 
 // Information needed to construct execution providers.
 struct VitisAIExecutionProviderInfo {
-  int device_id{0};
-  std::string backend_type;
-  std::string export_runtime_module;
-  std::string load_runtime_module;
+  VitisAIExecutionProviderInfo(const char* opt_str) : opt_str_(opt_str){};
+  const char* opt_str_;
```
[the remark this answers is not in the document]
Comment (wcy123) [reply onnxruntime/core/providers/vitisai/vitisai_execution_provider.h]: Yes, we are going to submit another PR for document very soon. It is under internal review for now.

File: include/onnxruntime/core/providers/vitisai/vitisai_provider_factory.h
[the remark this answers is not in the document]
Comment (wcy123) [reply include/onnxruntime/core/providers/vitisai/vitisai_provider_factory.h:1]: Yes. Otherwise we will get a compilation error, because https://github.com/microsoft/onnxruntime/blob/fe1ddd7b61c5c715f9a4678b994c844819d90c47/tools/ci_build/gen_def.py#L71

L71 generates a statement `#include <core/providers/vitisai/vitisai_provider_factory.h>`

File: onnxruntime/core/providers/vitisai/vitisai_execution_provider.cc (hunk 7 above)
Comment (wcy123) [reply onnxruntime/core/providers/vitisai/vitisai_execution_provider.cc:41]: In the future, we will put the VitisAI EP into an independent DLL which can be separately compiled.

We don't know whether there are alternative internal ORT APIs. Do you have any suggestion so that we can update accordingly.

File: include/onnxruntime/core/providers/vitisai/vitisai_provider_factory.h
[the remark this answers is not in the document]
Comment (wcy123) [reply include/onnxruntime/core/providers/vitisai/vitisai_provider_factory.h:1]: In 15d2c045bc1bf317f0b030975d4dbbdab23fef14, I put `vitisai` into the blacklist like others, and `vitisai_provider_factory.h` is not longer needed.
